### ISSUE TITLE
2FA

### ISSUE THREAD

**Issue by UnrealUNS** [keybase/keybase-issues#1522]
I don't know if this has been suggested yet, but two-factor authentication should definitely be implemented.

**Comment by zQueal** [comment]
This is already on the roadmap. Authy has been discussed extensively.

**Comment by callesoe** [comment]
This is absolutely also at the top of my wish list for keybase.

**Comment by zQueal** [comment]
Agreed--same for me. I'm a huge supporter of 2FA--however, I doubt we'll see this in the Node version. Currently the Devs are working on a Go version of Keybase which is gonna be a lot faster and hopefully easier to work with. Once the migration is done with security is feasibly the most important thing after the switch.

**Comment by sebie** [comment]
Will other 2FA methods be supported to use or just one, Authy?

**Comment by 5nafu** [comment]
Would you consider the [HOTP](https://en.wikipedia.org/w/index.php?title=HOTP) ([rfc4226](https://tools.ietf.org/html/rfc4226)) or [TOTP](https://en.wikipedia.org/wiki/Time-based_One-time_Password_Algorithm) ([rfc6238](https://tools.ietf.org/html/rfc6238)) standard instead of a proprietary solution? 
This way you are independent from a 3rd party and users can use the client they trust most (FreeOTP, Google Authenticator, RSA SecureID, ...)

**Comment by rminnick** [comment]
+1 for 2FA - I'm not a fan that when testing the KBFS my end-to-end encryption and signing is only as strong as my Keybase.io password.

**Comment by zQueal** [comment]
It actually seems that [almost a year ago](https://github.com/keybase/keybase-issues/wiki/A-Plan-for-Two-Factor-Auth) Keybase devs decided that traditional 2FA isn't what they were shooting for. [Original post](https://github.com/keybase/keybase-issues/issues/808#issuecomment-123740368).

**Comment by rminnick** [comment]
@zQueal makes a huge amount of sense, but until implemented as described, it's still a problem for anyone keeping all of the eggs in the online-compatible basket.  

My solution has just been to keep my private key out of the Keybase ecosystem, which limits the amount of damage a bad-actor could do (for example, they can't sign, encrypt, track, or untrack from the web - but of course, neither can I).

**Comment by zQueal** [comment]
I'd say that until some semblance of 2FA is employed keeping your private key out of the Keybase infrastructure is just a plain good idea. KBFS seems to be the primary focus right now--and it doesn't seem to be an easy task but hopefully we'll get 2FA very soon so we can all rest a little easier.

**Comment by vwal** [comment]
As an alternative to GA-compatible Authy it would be great to see [Duo](https://duo.com/) integration as well. Not much has happened in this thread for over a year, and it would be great to see 2fa in some form (whether it is one or more of the "standard" ways to implement 2fa, or the proprietary approach discussed in the [2016 document](https://github.com/keybase/keybase-issues/wiki/A-Plan-for-Two-Factor-Auth)).

**Comment by Morthawt** [comment]
U2F (Universal 2nd Factor, it's not regular "two factor" acronym) 2 factor would be a very big boost to this service, as long as you can choose backups like TOTP and backup codes and things.

**Comment by symgryph** [comment]
Add me to the 'might be better than slack, but still no go for security w/o 2fa!' list list.

**Comment by maxtaco** [comment]
We are about to launch a feature called “lockdown mode” which is what people who want 2FA need in the case of keybase. Details coming soon.

**Comment by maxtaco** [comment]
(teaser: and no need to copy 6-digit codes around...)

**Comment by editedbaseline** [comment]
Is it linked to [this](https://github.com/keybase/keybase-issues/issues/808#issuecomment-123740368
) from 2015?

Without prejudicing the "lockdown mode" feature before it comes out, if it doesn't offer equivalent or greater security than 2FA, then I'll be walking from Keybase. The security of the login can no longer be called sufficient to be able to take the rest of a security product seriously.

Also, will "lockdown mode" be proprietary or will it be available, preferably license free, to other organisations for them to use as they see fit on their sites/in their environments? If not, this will be yet another thing to do/to carry/to control/to look after/etc on top of Google Authenticator, physical tokens, a phone for SMS codes, etc

**Comment by maxtaco** [comment]
It's actually live now, so I can say. It's simply that if you opt into this account setting, the Website becomes roughly read-only. You need a Keybase provisioned, unlocked device to make any changes to your account.  We argue that this provides almost all of the properties of 2FA that people want in the context of Keybase (and it says nothing about other applications of 2FA).  If you get phished, or someone owns your email account, or someone brute-forces your passphrase, they can't reset your account, download encrypted keys, or delete your account, etc.  They'd need your physical device (or root access to your physical device) to do that.

For those on Linux, you can check it out with today's build: `keybase account lockdown`.  In the next release, it'll be in all of the apps and platforms.  If you opt into lockdown on one of your devices, all devices are affected.

**Comment by vwal** [comment]
If you're in lockdown mode and lose all your devices/access to them, but have your paper key/chief mask still available, is it possible to recover access at that point?

And, is there/will there be any way for the lockdown mode to be required by a third party? I'm guessing not, but let's say we require 2FA to be enabled by policy in an organization and would like to do the same for Keybase users in the company context (i.e. "in order for the company secrets to be transited over Keybase, then the keybase mode must be enabled").

**Comment by maxtaco** [comment]
> If you're in lockdown mode and lose all your devices/access to them, but have your paper key/chief mask still available, is it possible to recover access at that point?

Yes.

> And, is there/will there be any way for the lockdown mode to be required by a third party? I'm guessing not, but let's say we require 2FA to be enabled by policy in an organization and would like to do the same for Keybase users in the company context (i.e. "in order for the company secrets to be transited over Keybase, then the keybase mode must be enabled").

I'm nervous about the privacy angle here. You might not want other people to know what option you've picked here. Even without lockdown mode, someone who guesses your login password and had Web access to your account, but without device access, would not have access to any encrypted files or chats, since he'd also need a device key to decrypt it.

**Comment by junderw** [comment]
How can I activate lockdown mode?

**Comment by SoreGums** [comment]
so if using lockdown - someone gets a hold of a device. they can still just open the app and access everything. Like I'm at a certain chcekpoint and someone is like `so, you want to keep going - unlock this screen` you kind of have to do that. then they just open keybase and `all your base are belong to us`... however with a 2nd factor, you then need to provide that as well, and if you had two setup, 1 that goes to the `everyone can read this`, 2 that goes to the real account, things are a lot more secure. ofc you could argue well if you don't keep activity up in the diversion account that is going to fail, sure - that's not a reason to not provide this feature/functionality...

If the person then is like, I don't believe you, give me the real 2fa code - you have plausible deniability, they asked for the code, you gave it and the account unlocked - what more can they do... pretty sure by this point they need to formally charge you as a terrorist and then go through that paperwork...

security products or privacy-focused products always seem to leave out the coercion aspect when it comes to actually secure a product. sure it is more work, however, if the goal is privacy and security, it either is or it isn't. and this forced compliance aspect is real and ever more invoked these days.

maybe keybase can't be 2fa'd effectively - that hasn't really been said in this thread though. once a device is authenticated if the device goes out of user-control it is on the user to revoke that device. i believe there really needs to be an application level barrier rather than "well if you give your device to someone and they tell you to unlock it, that's your fault, don't do that".

**Comment by maxtaco** [comment]
For most people the 2FA app is also on the phone so having the unlocked
phone gives all factors.  I have a story for what to do when you know your
phone is going to get imaged at a security checkpoint. You can write me
privately if you’d like.

99% of the benefit of 2FA is against remote attackers. In person attacks
are way more difficult to defend against.

**Comment by SoreGums** [comment]
ofc - however 2fa means "2nd factor" it doesn't mean a code generated in an app on your phone. it could be a memorised 12-word string. I've memorised a few of them for different purposes. the diversion account could use the local app TOTP and then the real one is in your head. likewise the local TOTP app could also have the diversion system in place...

If someone needed to do things at this level on purpose, I don't have a need it's just something that I have observed when it comes to these kinds of products, then products like keybase really help these people as long as they don't have to wipe everything they carry for periods of time.

**Comment by junderw** [comment]
The problem with 2FA on keybase is that adding it as just a simple login extra step is misleading.

Once you get the private keys, logging in is unnecessary. You can just decrypt all the data.

This current model doesn't add flimsy "security theater" because it will lead to people leaving their devices unlocked and lending devices with their private keys to friends.

Physical attacks are difficult to defend against as your physical attack risk factors depend highly on what device you use and the security features of that device.

A 6 digit code as an "open sesame" to the server session key will not help anyone that isn't already helped by lockdown mode.
